polygon/p2p/message_listener.go

```diff
@@ -0,0 +1,141 @@
+	UnregisterBlockHeaders66Observer(observer MessageObserver[*sentry.InboundMessage])
+}
+
+func NewMessageListener(logger log.Logger, sentryClient direct.SentryClient) MessageListener {
+	return &messageListener{
+		logger:                  logger,
+		sentryClient:            sentryClient,
+		inboundMessageObservers: map[sentry.MessageId]map[MessageObserver[*sentry.InboundMessage]]struct{}{},
+	}
+}
+
+type messageListener struct {
+	once                    sync.Once
+	streamCtx               context.Context
+	streamCtxCancel         context.CancelFunc
+	logger                  log.Logger
+	sentryClient            direct.SentryClient
+	observersMu             sync.Mutex
+	inboundMessageObservers map[sentry.MessageId]map[MessageObserver[*sentry.InboundMessage]]struct{}
+	stopWg                  sync.WaitGroup
+}
+
+func (ml *messageListener) Start(ctx context.Context) {
+	ml.once.Do(func() {
+		ml.streamCtx, ml.streamCtxCancel = context.WithCancel(ctx)
+		go ml.listenBlockHeaders66()
+	})
+}
+
+func (ml *messageListener) Stop() {
+	ml.streamCtxCancel()
+	ml.stopWg.Wait()
+}
+
+func (ml *messageListener) RegisterBlockHeaders66Observer(observer MessageObserver[*sentry.InboundMessage]) {
+	ml.registerInboundMessageObserver(observer, sentry.MessageId_BLOCK_HEADERS_66)
+}
+
+func (ml *messageListener) UnregisterBlockHeaders66Observer(observer MessageObserver[*sentry.InboundMessage]) {
+	ml.unregisterInboundMessageObserver(observer, sentry.MessageId_BLOCK_HEADERS_66)
+}
+
+func (ml *messageListener) registerInboundMessageObserver(observer MessageObserver[*sentry.InboundMessage], messageId sentry.MessageId) {
+	ml.observersMu.Lock()
+	defer ml.observersMu.Unlock()
+
+	if observers, ok := ml.inboundMessageObservers[messageId]; ok {
+		observers[observer] = struct{}{}
+	} else {
+		ml.inboundMessageObservers[messageId] = map[MessageObserver[*sentry.InboundMessage]]struct{}{
+			observer: {},
+		}
+	}
+}
+
+func (ml *messageListener) unregisterInboundMessageObserver(observer MessageObserver[*sentry.InboundMessage], messageId sentry.MessageId) {
+	ml.observersMu.Lock()
+	defer ml.observersMu.Unlock()
+
+	if observers, ok := ml.inboundMessageObservers[messageId]; ok {
+		delete(observers, observer)
+	}
+}
+
+func (ml *messageListener) listenBlockHeaders66() {
+	ml.listenInboundMessage("BlockHeaders66", sentry.MessageId_BLOCK_HEADERS_66)
+}
+
+func (ml *messageListener) listenInboundMessage(name string, msgId sentry.MessageId) {
+	ml.stopWg.Add(1)
+	defer ml.stopWg.Done()
+
+	sentrymulticlient.SentryReconnectAndPumpStreamLoop(
+		ml.streamCtx,
+		ml.sentryClient,
+		ml.statusDataFactory(),
+		name,
+		ml.messageStreamFactory([]sentry.MessageId{msgId}),
+		ml.inboundMessageFactory(),
+		ml.inboundMessageHandler(),
+		nil,
+		ml.logger,
+	)
+}
+
+func (ml *messageListener) statusDataFactory() sentrymulticlient.StatusDataFactory {
+	return func() *sentry.StatusData {
+		return &sentry.StatusData{}
```

Comment:
TODO reminder

Reply:
I actually dont think updating the status should happen in the message listener - I have plans to add a new component for updating our status but that is a work for a future task